### PR TITLE
Feature/sc 67467/proper error handling for payments php and

### PR DESCRIPTION
**Pull request by abrahambosch**
added custom exceptions around order creation, payment, invoice creation, invoice email, and payment method saving. 
added logging to our customer logger for exceptions triggered within createPayment()

https://app.shortcut.com/publicsq/story/67467/proper-error-handling-for-payments-php-and-code-audit

### CODE REVIEW

**Review by rlfrahm — CHANGES_REQUESTED** [review]

File: PublicSquare/Payments/Model/Api/Payments.php [hunk 1]
```diff
@@ -179,151 +191,162 @@ public function __construct(
...
+            return $result;
+        } catch (\Exception $e) {
+            $this->logger->error($e->getMessage(), ['trace' => $e->getTraceAsString()]);
+            throw new \Magento\Framework\Exception\PaymentException(new \Magento\Framework\Phrase($e->getMessage()));
         }
-        return $result;
     } //end createApplication()
 
-    private function invoiceOrder(
+    private
```
[the remark this answers is not in the document]
Comment (abrahambosch) [reply PublicSquare/Payments/Model/Api/Payments.php]: `private` and `function` should be on the same line. I don't know how that happened.

File: PublicSquare/Payments/Model/Api/Payments.php [hunk 2]
```diff
@@ -396,31 +431,38 @@ public function createTransaction($order = null, $paymentData = array())
             $payment->setParentTransactionId(null);
             $payment->save();
             $order->save();
-            return  $transaction->save()->getTransactionId();
+            return $transaction->save()->getTransactionId();
         } catch (Exception $e) {
-            //log errors here
+            // log errors here? .. logging handled in parent class
+            // $this->logger->error($e->getMessage(), ['trace' => $e->getTraceAsString()]);
```
Comment (rlfrahm) [inline PublicSquare/Payments/Model/Api/Payments.php:437]: Remove these comments

File: PublicSquare/Payments/Model/Api/Payments.php [hunk 3]
```diff
@@ -335,23 +367,26 @@ private function invoiceOrder(
...
+    {
+        try {
             $this->invoiceSender->send($invoice);
             return true;
-        }
-        catch (\Exception $e)
-        {
+        } catch (\Exception $e) {
+            $this->logger->error($e->getMessage(), ['trace' => $e->getTraceAsString()]);
             // $this->logError($e->getMessage(), $e->getTraceAsString());
```
Comment (rlfrahm) [inline PublicSquare/Payments/Model/Api/Payments.php:390]: Remove this comment

File: PublicSquare/Payments/Model/Api/Payments.php [hunk 4]
```diff
@@ -179,151 +188,174 @@ public function __construct(
...
+            if ($orderId) {
+                $result["order_id"] = $orderId;
+            } else {
+                $result = ["error" => 1, "msg" => "Your custom message"];
+            }
+            return $result;
+        } catch (\Exception $e) {
+            $this->logger->error($e->getMessage(), ['trace' => $e->getTraceAsString()]);
+            throw $e;
+            //throw new \Magento\Framework\Exception\PaymentException(new \Magento\Framework\Phrase($e->getMessage()));
```
Comment (rlfrahm) [inline PublicSquare/Payments/Model/Api/Payments.php:349]: Remove this comment

**Review by rlfrahm — CHANGES_REQUESTED** [review]

**Review by rlfrahm — APPROVED** [review]

**Pull request merged**